doc/designs/quic-design/quic-io-arch.md

```diff
@@ -0,0 +1,483 @@
+
+- It is quite likely that there are buggy OSes out there which perform spurious
+  wakeups from select(3).
+
+- The fact that a socket is writeable does not necessarily mean that a datagram
+  of the size we wish to send is writeable, so a send(3) call could block
+  anyway.
+
+- This usage pattern precludes multithreaded use barring some locking scheme
+  due to the possibility of other threads racing between the call to select(3)
+  and the subsequent I/O call. This undermines our intentions to support
+  multi-threaded network I/O on the backend.
+
+Moreover, our QUIC implementation will not drive the Berkeley sockets API
+directly but uses the BIO abstraction to access the network, so these issues are
+then compounded by the limitations of our existing BIO interfaces. We do not
+have a BIO interface which provides for select(3)-like functionality or which
+can implement the required semantics above. Therefore, trying to implement QUIC
+on top of blocking I/O in this way would require violating the BIO abstraction
+layer, and would not work with custom BIOs.
+
+### Blocking sockets and threads
+
+Another conceptual possibility is that blocking calls could be kept ongoing in
+parallel threads. Under this model, there would be three threads:
+
+- a thread which exists solely to execute blocking calls to the `BIO_write` of
+  an underlying network BIO,
+- a thread which exists solely to execute blocking calls to the `BIO_read` of an
+  underlying network BIO,
+- a thread which exists solely to wait for and dispatch timeout events.
+
+This has a large number of disadvantages:
+
+- There is a hard requirement for threading functionality in order to be
+  able to support blocking semantics at the application level. Use of blocking
+  semantics at the application level will have a hard requirement on use of the
+  thread assisted mode. In environments where threading support is not available
+  or desired, our APIs will only be usable in a non-blocking fashion.
+
+- Several threads are spawned which the application is not in control of.
+  This undermines our general approach of providing the application with control
+  over OpenSSL's use of resources, such as allowing the application to do its
+  own polling or provide its own allocators.
+
+  At a minimum for a client, there must be two threads per connection. This
+  means if an application opens many outgoing connections, there will need
+  to be `2n` extra threads spawned.
+
+- By blocking in `BIO_write` calls, this precludes correct implementation of
+  QUIC. Unlike any analogue in TLS, QUIC packets are time sensitive and intended
+  to be transmitted as soon as they are generated. QUIC packets contain fields
+  such as the ACK Delay value, which is intended to describe the time between a
+  packet being received and a return packet being generated. Correct calculation
+  of this field is necessary to correct calculation of connection RTT. It is
```

Comment:
QUIC packets don't contain the time they were sent. They, or more specifically ACK frames, contain the amount of time *between* the time at which a packet X was received, and the time at which the ACK frame acknowledging that packet X was sent. This value is used to estimate connection RTT between the two endpoints.

In other words:
1. Host A sends packet X to Host B, remembering that it sent it at time T1
2. Host B waits for t ms before deciding to send an ACK
3. Host B generates packet Y to send to Host A, which acknowledges packet X and specifies that it waited t ms before sending the ACK
4. Packet Y is sent.
5. Host A receives packet Y at time T2
6. Host A estimates the connection RTT as `T2 - T1 - t`

If (4) blocks and the transmission of the packet is therefore delayed, then the calculation of `t` is inaccurate and RTT is not correctly estimated. It is a requirement for QUIC that the packet only be generated when it can be sent.

Reply:
So does this mean we want to delay filling in `t` as long as possible? Is the value of `t` signed?

Reply:
t is inside the encrypted packet, so it is authenticated against forgery. Thus the latest it can be generated is when we generate the packet. And yes, we want therefore to delay packet generation until we are ready to send.